homeassistant/components/http/forwarded.py

```diff
@@ -0,0 +1,174 @@
+
+    Process IP addresses, proto and host information in the forwarded for headers.
+
+    `X-Forwarded-For: <client>, <proxy1>, <proxy2>`
+    e.g., `X-Forwarded-For: 203.0.113.195, 70.41.3.18, 150.172.238.178`
+
+    We go through the list from the right side, and skip all entries that are in our
+    trusted proxies list. The first non-trusted IP is used as the client IP. If all
+    items in the X-Forwarded-For are trusted, including the most left item (client),
+    the most left item is used. In the latter case, the client connection originated
+    from an IP that is also listed as a trusted proxy IP or network.
+
+    `X-Forwarded-Proto: <client>, <proxy1>, <proxy2>`
+    e.g., `X-Forwarded-Proto: https, http, http`
+    OR `X-Forwarded-Proto: https` (one entry, even with multiple proxies)
+
+    The X-Forwarded-Proto is determined based on the corresponding entry of the
+    X-Forwarded-For header that is used/chosen as the client IP. However,
+    some proxies, for example, Kubernetes NGINX ingress, only retain one element
+    in the X-Forwarded-Proto header. In that case, we'll just use what we have.
+
+    `X-Forwarded-Host: <host>`
+    e.g., `X-Forwarded-Host: example.com`
+
+    If the previous headers are processed successfully, and the X-Forwarded-Host is
+    present, it will be used.
+
+    Additionally:
+      - If no X-Forwarded-For header is found, the processing of all headers is skipped.
+      - Log a warning when untrusted connected peer provides X-Forwarded-For headers.
+      - If multiple instances of X-Forwarded-For, X-Forwarded-Proto or
+        X-Forwarded-Host are found, an HTTP 400 status code is thrown.
+      - If malformed or invalid (IP) data in X-Forwarded-For header is found,
+        an HTTP 400 status code is thrown.
+      - The connected client peer on the socket of the incoming connection,
+        must be trusted for any processing to take place.
+      - If the number of elements in X-Forwarded-Proto does not equal 1 or
+        is equal to the number of elements in X-Forwarded-For, an HTTP 400
+        status code is thrown.
+      - If an empty X-Forwarded-Host is provided, an HTTP 400 status code is thrown.
+      - If an empty X-Forwarded-Proto is provided, or an empty element in the list,
+        an HTTP 400 status code is thrown.
+    """
+
+    @middleware
+    async def forwarded_middleware(request, handler):
+        """Process forwarded data by a reverse proxy."""
+        overrides = {}
+
+        # Handle X-Forwarded-For
+        forwarded_for_headers = request.headers.getall(X_FORWARDED_FOR, [])
+        if not forwarded_for_headers:
+            # No forwarding headers, continue as normal
+            return await handler(request)
+
+        # Ensure the IP of the connected peer is trusted
+        connected_ip = ip_address(request.transport.get_extra_info("peername")[0])
+        if not any(connected_ip in trusted_proxy for trusted_proxy in trusted_proxies):
```

Comment:
```suggestion
        if not any(connected_ip == trusted_proxy for trusted_proxy in trusted_proxies):
```
I think explicit is better here

Reply:
Maybe we should just look at the latest proxy in the list? We can not know the trust path

Reply:
I agree. It is implemented backward compatible (this is how the previous implementation did it as well). Doing this, might be a breaking change for some.

However, I agree. So maybe we should just go and do this?

Reply:
The `connected_ip` is a `ip_address` instance while the `trusted_proxy` is a `ip_network` instance. A network can contain more than one ip address. That's why there's a membership test, I think.

Reply:
That is correct @MartinHjelmare, almost forgot about that. Thanks 👍